### ISSUE TITLE
ServiceWorkerMessageEvent isn't consistent with MessageEvent

### ISSUE THREAD

**Issue by smaug----** [w3c/ServiceWorker#989]
Because of https://github.com/whatwg/html/issues/1882, the interfaces have now different behavior when handling .ports for example.

But I'd recommend just merge the interfaces.

**Comment by smaug----** [comment]
https://github.com/w3c/ServiceWorker/issues/988 is also related

**Comment by cdumez** [comment]
This also impacts ExtendableMessageEvent:
https://w3c.github.io/ServiceWorker/#extendablemessage-event-section

MessageEvent's ports attribute is no longer nullable in the HTML specification. It'd be great to align the similar interfaces in the Service Worker specification (or even merge the interfaces as @smaug----  suggests).

**Comment by foolip** [comment]
Ping @jungkees (most active committer recently)

**Comment by annevk** [comment]
@jungkees this is great, but there was also the suggestion to merge MessageEvent and ServiceWorkerMessageEvent. That's still pretty doable I think, no?

**Comment by jungkees** [comment]
Having looked at the [previous discussion around this](https://github.com/w3c/ServiceWorker/issues/669), we wanted the same for `ServiceWorkerMessageEvent` at the beginning. Later, for `ExtendableMessageEvent`, we had a reason not having been able to extend `MessageEvent`. I think even if we merge `ServiceWorkerMessageEvent` and `MessageEvent` at this time, we anyway have to maintain `ExtendableMessageEvent` interface separately as it's not a `MessageEvent`. Considering they have been shipped for a while, I'm inclined to regard them as pair of events defined in SW spec.

/cc @jakearchibald

**Comment by annevk** [comment]
One less class seems good? And `MessageEvent` / `ExtendableMessageEvent` we could potentially migrate to a shared mixin at some point.

**Comment by jungkees** [comment]
I'm not against refactoring. I've been concerned about the impact of the changes to the existing codes. If we do this, implementations should change, but the application codes won't be affected, right?

**Comment by annevk** [comment]
They could be affected if they use the class in some way. Extend its prototype, typeof it, etc. Seems unlikely, but there's a risk.

**Comment by annevk** [comment]
To be clear, that's only a problem for `ServiceWorkerMessageEvent`. The other change would not be noticeable (and implementations wouldn't even have to copy us necessarily there if they didn't care for it).

**Comment by foolip** [comment]
11 hits for "ServiceWorkerMessageEvent" results from httparchive:har.2016_09_15_chrome_requests_bodies and 6 from httparchive:har.2016_09_15_android_requests_bodies. They're things like http://docs.devdocs.io/dom/index.json?1469400308 and no real usage that I can see.

But, the Blink change would have to happen soon, if the spec changes but Blink waits, we could end up in a very silly place.

**Comment by jungkees** [comment]
> But, the Blink change would have to happen soon, if the spec changes but Blink waits, we could end up in a very silly place.

Indeed. Other browsers also won't be able to completely avoid being in this kind of silly place for the time being I think.

**Comment by annevk** [comment]
@foolip are you okay with trying/doing it? Should be easy enough for others to follow if so.

**Comment by foolip** [comment]
I see it's also in Gecko. If one of Blink and Gecko changes right away and the other commits to following before long, I think the risk looks incredibly low, so :+1:

**Comment by jakearchibald** [comment]
I'm happy with this changing. Highly unlikely to break anything, and hopefully we'll see it in beta etc if it does.

**Comment by foolip** [comment]
So I guess step 1 would be to update MessageEvent in HTML to use `(WindowProxy or MessagePort or ServiceWorker)`?

**Comment by jungkees** [comment]
Considering the shared mixin for `MessageEvent` and `ExtendableMessageEvent`, wouldn't it be `(WindowProxy or MessagePort or ServiceWorker or Client)`?

**Comment by foolip** [comment]
I wasn't thinking about `ExtendableMessageEvent`, but yeah, I guess it's just as well to fix it all at once. @annevk, do you feel like trying the HTML changes?

**Comment by annevk** [comment]
For mixins:

* I think I'd rather wait until IDL has actual mixins
* The other issue is what to do with the dictionary, since dictionaries don't support mixins, we'd have to duplicate there for now

Given that, I think we should postpone deduplication there, especially as it doesn't directly impact implementations or developers, but I'll raise an IDL issue on mixins for dictionaries.

For removing `ServiceWorkerMessageEvent`, I can make the change to HTML to make `MessageEvent` reusable for that case.

**Comment by foolip** [comment]
SGTM.

**Comment by jungkees** [comment]
Sounds good.

**Comment by annevk** [comment]
Thanks, done.

**Comment by LJWatson** [comment]
Thanks Remy. Red wine for me... or tequila of course :)

Léonie.
-- 
@LeonieWatson tink.uk Carpe diem

**Comment by annevk** [comment]
@jungkees I think it would be better if the generated results are somewhat wrong or something is not working well to open a PR and get those issues resolved first. Doing it through comments on a commit is very hard to follow.

**Comment by jungkees** [comment]
Okay. Got it.

**Comment by annevk** [comment]
That also makes the review process easier if we start doing more through PRs here. It's been working very well over in WHATWG-land.

**Comment by jungkees** [comment]
Understood. I'll try to do so. Thanks!
